java/org/contikios/cooja/Simulation.java

```diff
@@ -354,36 +396,42 @@ public void stopSimulation() {
    * @param rv Return value from startSimulation, should be null unless > 0
    */
   public void stopSimulation(boolean block, Integer rv) {
-    if (!isRunning()) {
+    if (!isRunning() || isShutdown) {
       return;
     }
     assert rv == null || rv > 0 : "Pass in rv = null or rv > 0";
     if (rv != null) {
       returnValue = rv;
     }
-    stopSimulation = true;
 
-    if (block) {
-      if (Thread.currentThread() == simulationThread) {
-        return;
-      }
+    commandQueue.add(STOP);
 
-      /* Wait until simulation stops */
-      try {
-        Thread simThread = simulationThread;
-        if (simThread != null) {
-          simThread.join(100);
-        }
-      } catch (InterruptedException e) {
-      }
+    if (block) {
+      waitFor(false, 250);
     }
-    cooja.updateProgress(true);
   }
 
-  public void stopScriptEngines() {
-    // FIXME: this should be at the end of run(), but that requires the sim thread to be persistent.
-    for (var engine : scriptEngines) {
-      engine.deactivateScript();
+  private void waitFor(boolean isRunning, long maxTime) {
+    if (Thread.currentThread() == simulationThread) {
+      return;
+    }
+    long startTime = System.currentTimeMillis();
+    try {
+      synchronized (stateLock) {
+        while (this.isRunning != isRunning && !isShutdown) {
+          if (maxTime == 0) {
```

Comment:
I think the loop body can be replaced with:
```Java
long waitTime = maxTime - (System.currentTimeMillis() - startTime);
if (maxTime != 0 && waitTime <= 0) {
  return;
}
stateLock.wait(maxTime == 0 ? 0 : waitTime);
```
(And waitTime should probably be called `timeout`)

Reply:
Updated. Thanks!